### ISSUE TITLE
[no-unused-vars] v4 false positive with generic inference - with isolated reproduction repo

### ISSUE THREAD

**Issue by willheslam** [typescript-eslint/typescript-eslint#2459]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
https://github.com/willheslam/typescript-unused-var-reprod
`npm run lint`

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": 1
  }
}
```

```TS
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never
```

**Expected Result**

No linting issues

**Actual Result**

```
3:12  warning  'I' is defined but never used  @typescript-eslint/no-unused-vars

✖ 1 problem (0 errors, 1 warning)
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `4.0.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.0` |
| `node`                             | `12.14.1` |

**Comment by bradennapier** [comment]
same false positive happens if you use `?.`

```typescript
e.outputMessages.map((m) => ({
          type: m?.type,
          code: (m as any)?.code,
        })),
```

![image](https://user-images.githubusercontent.com/15365418/92019345-77a30780-ed24-11ea-8c76-71789a3b710b.png)

**Comment by bradzacher** [comment]
please raise a separate issue. That issue is unrelated to this one.

**Comment by alecf** [comment]
I have a related fail, haven't narrowed it down, but it looks something like this:
```typescript
export interface KeyedListActionCreatorsWithExtra<
  T,
  EA extends any[],
  EP,
  K = string
> {
  setList: (key: K, list: T[], ...extra: EA) => SetKeyedListAction<T, K> & EP;
  ...
```

The claim is that `EA` is unused - so it might not be the generic inference here...

**Comment by alecf** [comment]
Ah, it's the `...args: EA`. Repro'ed here:

https://github.com/alecf/typescript-unused-var-reprod

LMK if I should file another issue, it seems like it is related to this one but not sure if the fix would be related or not

**Comment by bradzacher** [comment]
@alecf - there's no related issues here. Each issue is isolated and separate.
Your issue is already reported here: https://github.com/typescript-eslint/typescript-eslint/issues/2449

**Comment by blair** [comment]
I'm seeing a similar warning in this code:

```
// https://stackoverflow.com/a/51955852/
type KnownKeys<T> = {
    [K in keyof T]: string extends K ? never : number extends K ? never : K
} extends { [_ in keyof T]: infer U } ? U : never;
```

gives

```
  4:35  error  'U' is defined but never used  @typescript-eslint/no-unused-vars
```

**Comment by StanislavNemytov** [comment]
```export type InferActionsTypes<T> = T extends { [keys: string]: (...args: any[]) => infer U } ? U : never;```
How about this. Help!
